### PR TITLE
Fiducial Marker Detection

### PR DESCRIPTION
**Pull request by trickeydan**

### CODE REVIEW

**Comment by trickeydan** [comment]
I'm aware the test coverage is almost non-existent. I think that there are high priorities currently. We should still do it asap though, just not in this PR

**Comment by trickeydan** [comment]
Closes #301

File: j5/backends/hardware/zoloto/camera_board.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from j5.backends.hardware import HardwareEnvironment
+from j5.boards import Board
+from j5.boards.zoloto import ZolotoCameraBoard
+from j5.components import MarkerCameraInterface
+from j5.vision import Coordinate, Marker, MarkerList
+
+CAMERA_PATH = PosixPath("/dev/video0")
+CAMERA_SERIAL = "video0"
+
+T = TypeVar("T", bound=Camera)
```
Comment (HU90m) [inline j5/backends/hardware/zoloto/camera_board.py]: Is 'T' left over from development or does it serve a purpose?

Comment (trickeydan) [reply j5/backends/hardware/zoloto/camera_board.py]: It's used in the backend constructor to allow subclasses of Camera

**Comment by trickeydan** [comment]
Awaiting fix of upstream issue https://github.com/RealOrangeOne/zoloto/issues/41

File: j5/backends/console/zoloto/camera_board.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                Marker(marker_id, Coordinate(marker_x, marker_y, marker_z)),
+            )
+
+        return markers
+
+    def save_annotated_image(self, file: Path) -> None:
+        """Save an annotated image to a file."""
+        self._console.info(f"Saved annotated image to {file}")
+        self._console.info(
+            f"(Image not actually saved due to backend limitations.",
```
Comment (RealOrangeOne) [inline j5/backends/console/zoloto/camera_board.py]: Looks like you've got a stray bracket here, and this probably doesn't need to be an `f`-string

File: j5/backends/hardware/zoloto/camera_board.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+from zoloto.cameras.camera import Camera
+
+from j5.backends import Backend
+from j5.backends.hardware import HardwareEnvironment
+from j5.boards import Board
+from j5.boards.zoloto import ZolotoCameraBoard
+from j5.components import MarkerCameraInterface
+from j5.vision import Coordinate, Marker, MarkerList
+
+CAMERA_PATH = PosixPath("/dev/video0")
```
Comment (RealOrangeOne) [inline j5/backends/hardware/zoloto/camera_board.py]: Is there a reason this is `PosixPath` over simply `Path`?

File: j5/backends/hardware/zoloto/camera_board.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    board = ZolotoCameraBoard
+
+    @classmethod
+    def discover(cls, camera_class: Type[Camera] = DefaultCamera) -> Set[Board]:
+        """Discover boards that this backend can control."""
+        if system() != "Linux":
+            # We currently only support Zoloto on Linux platforms as there is
+            # no easy way to detect the presence of webcams on other platforms.
+            return set()
+        if not exists(CAMERA_PATH):
```
Comment (RealOrangeOne) [inline j5/backends/hardware/zoloto/camera_board.py]: You can simply call `CAMERA_PATH.exists()` here, which looks much nicer!

File: j5/backends/hardware/zoloto/camera_board.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @property
+    def firmware_version(self) -> Optional[str]:
+        """The firmware version reported by the board."""
+        return zoloto_version
+
+    def get_visible_markers(self, identifier: int) -> MarkerList:
+        """Get markers that are visible to the camera."""
+        markers = MarkerList()
+
+        marker_gen = self._zcamera.process_frame()
```
Comment (RealOrangeOne) [inline j5/backends/hardware/zoloto/camera_board.py]: I _highly_ recommend using `process_frame_eager` here, as it only does a pose estimation for each _size_ of marker, rather than for each marker it sees, which is a fair amount faster!

File: j5/boards/zoloto/camera_board.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    @property
+    def serial(self) -> str:
+        """Get the serial number."""
+        return self._serial
+
+    @property
+    def firmware_version(self) -> Optional[str]:
+        """Get the firmware version of the board."""
+        return None
```
Comment (RealOrangeOne) [inline j5/boards/zoloto/camera_board.py]: You can probably still use the version of `zoloto` here.

File: j5/boards/zoloto/camera_board.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    @property
+    def camera(self) -> MarkerCamera:
+        """The camera on this board."""
+        return self._camera
+
+    def make_safe(self) -> None:
+        """Make this board safe."""
+        pass
```
Comment (RealOrangeOne) [inline j5/boards/zoloto/camera_board.py]: Here you should call `.close` on the underlying zoloto camera

File: stubs/zoloto/__init__.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+"""Zoloto."""
+
+from enum import IntEnum
+
+__version__: str
+
+
+class MarkerDict(IntEnum):
```
Comment (RealOrangeOne) [inline stubs/zoloto/__init__.py]: Why is this necessary? The provided types should be more than enough

Comment (trickeydan) [reply stubs/zoloto/__init__.py]: Due to the way that mypy stubs work, it will completely ignore any types in your library unfortunately. The stubs can be removed at a later date when upstream has the hints and implements PEP 561

File: j5/backends/hardware/zoloto/camera_board.py (hunk 3 above)
Comment (trickeydan) [reply j5/backends/hardware/zoloto/camera_board.py]: Mostly because this is a unix path, but am happy to change.

File: j5/boards/zoloto/camera_board.py (hunk 6 above)
Comment (trickeydan) [reply j5/boards/zoloto/camera_board.py]: This is a bit of j5 which needs fixing, there is an issue for it #255. Have implemented the correct interim behaviour.

File: j5/backends/hardware/zoloto/camera_board.py (hunk 5 above)
Comment (trickeydan) [reply j5/backends/hardware/zoloto/camera_board.py]: I'm not bumping this until after the upstream is fixed
